src/models/data.cpp

```diff
@@ -292,6 +292,42 @@ TopDownResults TopDownResults::fromBottomUp(const BottomUpResults& bottomUpData)
     return results;
 }
 
+PerLibraryResults PerLibraryResults::fromTopDown(const TopDownResults& topDownData)
+{
+    PerLibraryResults results;
+    QHash<QString, int> binaryToResultIndex;
+    results.costs.initializeCostsFrom(topDownData.selfCosts);
+
+    std::function<void(const TopDown& parent)> topDownIterator = [&topDownIterator, &results, &binaryToResultIndex,
+                                                                  topDownData](const TopDown& parent) {
+        const auto binary = parent.symbol.binary;
+
+        if (!binaryToResultIndex.contains(binary)) {
```

Comment:
optimize for the common case that the binary will be found, you currently always do two hash lookups which is wasteful

```
auto resultIndexIt = binaryToResultIndex.find(binary);
if (resultIndexIt == binaryToResultIndex.end()) {
    resultIndexIt = binaryToResultIndex.insert(binary, binaryToResultIndex.size())
    ...
}

...
results.costs.add(*resultIndexIt);
```